src/main/kotlin/org/rust/ide/refactoring/changeSignature/impl.kt

```diff
@@ -0,0 +1,405 @@
+    }
+    renameParameters(factory, parameterRenames, allRenames)
+    changeAsync(factory, function, config)
+    changeUnsafe(factory, function, config)
+}
+
+private fun rename(factory: RsPsiFactory, function: RsFunction, config: RsChangeFunctionSignatureConfig) {
+    function.identifier.replace(factory.createIdentifier(config.name))
+}
+
+private fun renameFunctionUsage(
+    factory: RsPsiFactory,
+    usage: RsFunctionUsage,
+    config: RsChangeFunctionSignatureConfig
+) {
+    val identifier = factory.createIdentifier(config.name)
+    when (usage) {
+        is RsFunctionUsage.Reference -> usage.path.referenceNameElement?.replace(identifier)
+        is RsFunctionUsage.FunctionCall -> {
+            val path = (usage.call.expr as? RsPathExpr)?.path ?: return
+            path.referenceNameElement?.replace(identifier)
+        }
+        is RsFunctionUsage.MethodCall -> usage.call.identifier.replace(identifier)
+    }
+}
+
+fun changeVisibility(function: RsFunction, config: RsChangeFunctionSignatureConfig) {
+    if (function.vis?.text == config.visibility?.text) return
+
+    function.vis?.delete()
+
+    val vis = config.visibility
+    if (vis != null) {
+        function.addBefore(vis, function.firstChild)
+    }
+}
+
+private fun changeReturnType(factory: RsPsiFactory, function: RsFunction, config: RsChangeFunctionSignatureConfig) {
+    if (!areTypesEqual(function.retType?.typeReference, config.returnTypeReference)) {
+        function.retType?.delete()
+        if (config.returnType !is TyUnit) {
+            val ret = factory.createRetType(config.returnTypeReference.text)
+            function.addAfter(ret, function.valueParameterList) as RsRetType
+            RsImportHelper.importTypeReferencesFromTy(function, config.returnType, useAliases = true)
+        }
+    }
+}
+
+private fun changeParameters(
+    factory: RsPsiFactory,
+    function: RsFunction,
+    config: RsChangeFunctionSignatureConfig,
+    parameterOps: List<ParameterOperation>
+): Map<RsValueParameter, String> {
+    val parameters = function.valueParameterList ?: return emptyMap()
+    val parameterList = parameters.valueParameterList.toList()
+    val originalParameters = parameters.copy() as RsValueParameterList
+    val parameterRenames = mutableMapOf<RsValueParameter, String>()
+
+    cycle@ for ((index, data) in config.parameters.zip(parameterOps).withIndex()) {
+        val (parameter, op) = data
+        when (op) {
+            is ParameterOperation.Add -> {
+                deleteItem(parameterList, index)
+
+                val typeReference = parameter.typeReference
+                val newParameter = factory.createValueParameter(parameter.patText, typeReference, reference = false)
+                val anchor = findAnchorToInsertItem(parameters, function, index)
+                insertItemWithComma(factory, newParameter, parameters, anchor)
+                RsImportHelper.importTypeReferencesFromTy(function, parameter.type, useAliases = true)
```

Comment:
I tried that originally, but the import didn't work with the type reference, only with the type. Now that I have switched the parameter type to a String, the import doesn't work at all. For example, in this test:
```rust
mod foo {
    pub struct S;
             //^
}
mod bar {
    fn baz/*caret*/() {}
}
```
Adding a parameter with type `S` creates a type reference which has `type == TyUnknown`, so no import happens even if `importTypeReferencesFromElements` is used.

Reply:
> Now that I have switched the parameter type to a String, the import doesn't work at all.

Oh, so looks like we should switch back to `RsTypeReference` (but with rendering it as empty string if it is `null`)

> I tried that originally, but the import didn't work with the type reference, only with the type.

Actually I don't know how import can work, because if user accepts autocompletion, then import gets added even before refactoring was started, and otherwise type will indeed be unknown. Probably we should use some fresh file as context for type fields, and set original file as context for this fresh file (so any imports during dialog editing are added to fresh file and not original file)

https://user-images.githubusercontent.com/6505554/105155603-db836c80-5b1b-11eb-8ab9-56235c04ca5b.mp4

Reply:
> Oh, so looks like we should switch back to RsTypeReference (but with rendering it as empty string if it is null)

Well the type reference is only useful for tests, in GUI it will not refer to existing types anyway (unless they are auto-imported during editing). I'll try the separate file to see if it helps.

Reply:
The behaviour of importing something into a file from a code fragment is intended generally? Or should I change it for all code fragments?

Reply:
> The behaviour of importing something into a file from a code fragment is intended generally?

* For code fragment used in debugger definitely not intended - #6699
* For other code fragments I don't know, but I suppose that e.g. for [code fragment used in macro completion](https://github.com/intellij-rust/intellij-rust/blob/0f5790ecbd2d152fe9265d4b3160b9749269a48b/src/main/kotlin/org/rust/lang/core/completion/RsPartialMacroArgumentCompletionProvider.kt#L61-L64) it is intended behaviour

> Or should I change it for all code fragments?

So if possible maybe add some flag to code fragment constructor (to control behaviour case-by-case)

Reply:
Right, that makes sense, thanks. So far I only managed to change the import location from the file to the fragment itself :D Which is of course not ideal. I suppose that to create a virtual file, I'd need to somehow connect it to the corresponding module structure, so that it sees the correct paths from the perspective of the function, and then direct the imports into it.

I guess that e.g. the new macro engine also needs to build something like that? @vlad20012 any hints as to what API to use?

Reply:
We can try something similar to [used in repl console](https://github.com/intellij-rust/intellij-rust/blob/0f5790ecbd2d152fe9265d4b3160b9749269a48b/src/main/kotlin/org/rust/ide/console/RsConsoleCodeFragmentContext.kt#L103-L106) (I don't know whether it will work or not):
```kotlin
    val function: RsFunction = ...
    val freshFile = RsPsiFactory(project).createFile("fn main() {}")
    freshFile.originalFile = function.containingFile as RsFile
    return RsTypeReferenceCodeFragment(project, "", context = freshFile)
```

Reply:
Thanks for the hint, I included an `importTarget` in the code fragments and check it in `ImportCandidate.import`. It seems to work fine.

Reply:
Looks like it works in tests, but didn't work in IDE...

https://user-images.githubusercontent.com/6505554/105376387-bc1f3900-5c1a-11eb-9a5d-48b2c4598b14.mp4